gdnative-core/src/core_types/quat.rs

```diff
@@ -9,22 +12,248 @@ pub struct Quat {
+    #[inline]
+    pub fn length_squared(self) -> f32 {
+        self.glam().length_squared()
+    }
+
+    /// Returns a copy of the quaternion, normalized to unit length.
+    #[inline]
+    pub fn normalized(self) -> Self {
+        Self::gd(self.glam().normalize())
+    }
+
+    /// Sets the quaternion to a rotation which rotates around axis by the specified angle, in
+    /// radians. The axis must be a normalized vector.
+    #[inline]
+    pub fn set_axis_angle(&mut self, axis: Vector3, angle: f32) {
+        *self = Self::from_axis_angle(axis, angle)
+    }
+
+    /// Sets the quaternion to a rotation specified by Euler angles (in the YXZ convention: when
+    /// decomposing, first Z, then X, and Y last), given in the vector format as (X angle, Y angle,
+    /// Z angle).
+    #[inline]
+    pub fn set_euler(&mut self, euler: Vector3) {
+        *self = Self::from_euler(euler);
+    }
+
+    /// Returns the result of the spherical linear interpolation between this quaternion and to by
+    /// amount weight.
+    ///
+    /// **Note:** Both quaternions must be normalized.
+    #[inline]
+    pub fn slerp(self, b: Self, t: f32) -> Self {
+        debug_assert!(self.is_normalized(), "Quaternion `self` is not normalized");
+        debug_assert!(b.is_normalized(), "Quaternion `b` is not normalized");
+        Self::gd(self.glam().slerp(b.glam(), t))
+    }
+
+    /// Returns the result of the spherical linear interpolation between this quaternion and `t` by
+    /// amount `t`, but without checking if the rotation path is not bigger than 90 degrees.
+    #[inline]
+    pub fn slerpni(self, b: Self, t: f32) -> Self {
+        debug_assert!(self.is_normalized(), "Quaternion `self` is not normalized");
+        debug_assert!(b.is_normalized(), "Quaternion `b` is not normalized");
+        let dot = self.dot(b);
+        if dot.abs() > 0.9999 {
+            self
+        } else {
+            let theta = dot.acos();
+            let sin_t = 1.0 / theta.sin();
+            let new_f = (t * theta).sin() * sin_t;
+            let inv_f = ((1.0 - t) * theta).sin() * sin_t;
+            Self::new(
+                inv_f * self.x + new_f * b.x,
+                inv_f * self.y + new_f * b.y,
+                inv_f * self.z + new_f * b.z,
+                inv_f * self.w + new_f * b.w,
+            )
+        }
+    }
+
+    /// Returns a vector transformed (multiplied) by this quaternion.
+    #[inline]
+    pub fn xform(self, v: Vector3) -> Vector3 {
+        self * v
+    }
+
+    #[inline]
+    fn gd(quat: glam::Quat) -> Self {
+        Self::new(quat.x, quat.y, quat.z, quat.w)
+    }
+
     #[inline]
     fn glam(self) -> glam::Quat {
         glam::Quat::from_xyzw(self.x, self.y, self.z, self.w)
     }
 }
+
+impl Mul<Vector3> for Quat {
+    type Output = Vector3;
+
+    #[inline]
+    fn mul(self, with: Vector3) -> Vector3 {
+        debug_assert!(self.is_normalized(), "Quaternion is not normalized");
```

Comment:
Is normalization really required? Both Godot and glam documentation don't mention this.

Reply:
Godot throws an error if the quaternion isn't normalized:

```cpp
	_FORCE_INLINE_ Vector3 xform(const Vector3 &v) const {
#ifdef MATH_CHECKS
		ERR_FAIL_COND_V_MSG(!is_normalized(), v, "The quaternion must be normalized.");
#endif
		Vector3 u(x, y, z);
		Vector3 uv = u.cross(v);
		return v + ((uv * w) + u.cross(uv)) * ((real_t)2);
	}

```

Reply:
What if we just normalize it? I'd imagine that the cost of `is_normalized` would be very close to `normalized` itself.

Reply:
It is a `debug_assert`, so it won't be included in release builds and thus have no overhead.

Reply:
Sounds reasonable, for me we can leave the implementation like this.

I would document that that normalization is required though. Also for `xform()`.
And maybe mention that both `xform()` and `*` do the same thing.

Reply:
I just realized that Godot's `Quat operator*(const Vector3 &v)` doesn't call `xform` but instead looks like this:

```
	Quat operator*(const Vector3 &v) const {
		return Quat(w * v.x + y * v.z - z * v.y,
				w * v.y + z * v.x - x * v.z,
				w * v.z + x * v.y - y * v.x,
				-x * v.x - y * v.y - z * v.z);
	}
```

Glam's implementation says:

```
Multiplies a quaternion and a 3D vector, returning the rotated vector.
```

And it seems it indeed does:

```
   fn mul_vector3(self, other: XYZ<T>) -> XYZ<T> {
        glam_assert!(FloatVector4::is_normalized(self));
        let w = self.w;
        let b = XYZ {
            x: self.x,
            y: self.y,
            z: self.z,
        };
        let b2 = b.dot(b);
        other
            .mul_scalar(w * w - b2)
            .add(b.mul_scalar(other.dot(b) * T::TWO))
            .add(b.cross(other).mul_scalar(w * T::TWO))
    }
```

Reply:
So it seems that `Variant::evaluate` uses `xform` wen multiplying quaternions.

```
                        CASE_TYPE(math, OP_MULTIPLY, QUAT) {
                                switch (p_b.type) {
                                        case VECTOR3: {
                                                _RETURN(reinterpret_cast<const Quat *>(p_a._data._mem)->xform(*(const Vector3 *)p_b._data._mem));
                                        }
```

I guess we should keep using the `xform` implementation for `Mul<Vector3>` then to stay consistent with GDScript.

Reply:
> I just realized that Godot's `Quat operator*(const Vector3 &v)` doesn't call `xform` but instead looks like this:

How does `xform()` then look?

Do `quat * vector` and `quat.xform(vector)` do different things in GDScript?

Reply:
> Do `quat * vector` and `quat.xform(vector)` do different things in GDScript?

They do the same thing in GDScript.

Reply:
Thanks for checking!

So I'm OK with any implementation, as long as we reuse code (one method calls the other, and if it helps we can resort to glam). Would that work, or did I miss something?

Reply:
> Would that work, or did I miss something?

That'll work yes.